src/context/internal.rs

```diff
@@ -116,6 +117,37 @@ pub trait ContextInternal<'a>: Sized {
 
     fn activate(&self) { self.scope_metadata().active.set(true); }
     fn deactivate(&self) { self.scope_metadata().active.set(false); }
+
+    fn try_catch_internal<'b: 'a, F>(&mut self, f: F) -> Result<Handle<'a, JsValue>, Handle<'a, JsValue>>
+        where F: FnOnce(&mut Self) -> JsResult<'b, JsValue>
+    {
+        let p = Box::into_raw(Box::new(f)) as *mut c_void;
+        let mut local: raw::Local = unsafe { std::mem::zeroed() };
+        let threw = unsafe { neon_runtime::try_catch::with(try_catch_glue::<Self, F>, self as *mut Self as *mut c_void, p, &mut local) };
+        if threw {
+            Err(JsValue::new_internal(local))
+        } else {
+            Ok(JsValue::new_internal(local))
+        }
+    }
+}
+
+extern "C" fn try_catch_glue<'a, 'b: 'a, C: ContextInternal<'a>, F>(p: *mut c_void, cx: *mut c_void, ok: *mut bool, ok_value: *mut raw::Local)
+    where C: ContextInternal<'a>,
+          F: FnOnce(&mut C) -> JsResult<'b, JsValue>
+{
+    let f: F = *unsafe { Box::from_raw(p as *mut F) };
+
+    // FIXME: convert_panics(f)
```

Comment:
This is synchronous code, won't panics be caught by the `convert_panics` in `invoke` from the caller?

In my opinion, I don't think `try_catch` should be used to catch panics--which converting panics here would allow. I think it should only catch exceptions in V8.

tl;dr-- I don't think this is necessary or should be added.

Reply:
It doesn't seem obvious to me what the best behavior is here, but I think I'm fine with letting the panic keep unwinding all the way to the top level of the Neon function.

I think I should use `catch_unwind` and signal back to Rust to continue panicking though, right? I think you were saying that it's undefined behavior for Rust to panic across an FFI boundary?

Reply:
It was in the Unsafe Rust guide that I was reading. But, the comment might be dated. I've since found this:

https://github.com/rust-lang/rust/pull/65646

Reply:
This comment in the PR:

> I haven't updated any of the documentation, so officially unwinding through FFI is still UB. However this is a step towards making it well-defined.

suggests to me it's still not there yet. To be safe, I'll just catch any panics inside the inner Rust callback, return the boxed panic value across the FFI boundary, and re-panic from the outer `try_catch`. That way it avoids any undefined behavior but `try_catch` doesn't catch panics, as you suggest. I agree with your point about bad behavior -- people shouldn't be using Rust panics like exceptions and we shouldn't be encouraging it.